airbyte-cdk/java/airbyte-cdk/db-destinations/src/main/java/io/airbyte/cdk/integrations/destination/jdbc/typing_deduping/JdbcDestinationHandler.java

```diff
@@ -0,0 +1,119 @@
+    this.databaseName = databaseName;
+    this.jdbcDatabase = jdbcDatabase;
+  }
+
+  @Override
+  public Optional<TableDefinition> findExistingTable(final StreamId id) throws Exception {
+    return findExistingTable(jdbcDatabase, databaseName, id.finalNamespace(), id.finalName());
+  }
+
+  @Override
+  public boolean isFinalTableEmpty(final StreamId id) throws Exception {
+    final int rowCount = jdbcDatabase.queryInt(
+        """
+        SELECT row_count
+        FROM information_schema.tables
+        WHERE table_catalog = ?
+          AND table_schema = ?
+          AND table_name = ?
+        """,
+        databaseName,
+        id.finalNamespace(),
+        id.finalName());
+    return rowCount == 0;
+  }
+
+  @Override
+  public Optional<Instant> getMinTimestampForSync(final StreamId id) throws Exception {
+    return Optional.empty();
+  }
+
+  @Override
+  public void execute(final String sql) throws Exception {
+    if (sql == null || sql.isEmpty()) {
+      return;
+    }
+    final UUID queryId = UUID.randomUUID();
+    LOGGER.info("Executing sql {}: {}", queryId, sql);
+    final long startTime = System.currentTimeMillis();
+
+    try {
+      jdbcDatabase.execute(sql);
+    } catch (final SQLException e) {
+      LOGGER.error("Sql {} failed", queryId, e);
+      throw e;
+    }
+
+    LOGGER.info("Sql {} completed in {} ms", queryId, System.currentTimeMillis() - startTime);
+  }
+
+  public static Optional<TableDefinition> findExistingTable(final JdbcDatabase jdbcDatabase,
+                                                            final String databaseName,
+                                                            final String schemaName,
+                                                            final String tableName)
+      throws SQLException {
+    final DatabaseMetaData metaData = jdbcDatabase.getMetaData();
+    // TODO: normalize namespace and finalName strings to quoted-lowercase (as needed. Snowflake
+    // requires uppercase)
+    final LinkedHashMap<String, ColumnDefinition> columnDefinitions = new LinkedHashMap<>();
+    try (final ResultSet columns = metaData.getColumns(databaseName, schemaName, tableName, null)) {
+      while (columns.next()) {
+        final String columnName = columns.getString("COLUMN_NAME");
+        final String typeName = columns.getString("TYPE_NAME");
+        final int columnSize = columns.getInt("COLUMN_SIZE");
+        final int datatype = columns.getInt("DATA_TYPE");
+        SQLType sqlType;
+        try {
+          sqlType = JDBCType.valueOf(datatype);
+        } catch (final IllegalArgumentException e) {
+          // Unknown jdbcType convert to customSqlType
+          LOGGER.warn("Unrecognized JDBCType {}; falling back to UNKNOWN", datatype, e);
+          sqlType = new CustomSqlType("Unknown", "Unknown", datatype);
+        }
+        columnDefinitions.put(columnName, new ColumnDefinition(columnName, typeName, sqlType, columnSize));
```

Comment:
This was implemented before I went neck deep into Jooq.  Jooq has `create.meta()` which essentially retrieves JdbcMetadata in Jooq object formats.  I see in existingStreamMatchesTable we have to re-map them from SqlType to Jooq's DataType.  constructing this in DataType format, will that make it any easier

Reply:
Created https://github.com/airbytehq/airbyte/issues/33204